File: teams/shopify-edgescale/go/armstrong-numbers/armstrong_numbers.go
package armstrong

func digits(num int) int {
	d := 0

	for n := num; n > 0; n /= 10 {
		d++
	}

	return d
}

func pow(num int, power int) int {
	if power == 0 {
		return 1
	}
	return num * pow(num, power-1)
}

// IsNumber determines whether a number is an Armstrong number
func IsNumber(num int) bool {
	l := digits(num)

	s := 0

	for n := num; n > 0; n /= 10 {
		s += pow(n%10, l)
	}

	return s == num
}

File: clojure/armstrong-numbers/src/armstrong_numbers.clj
(ns armstrong-numbers)

(defn- digits [num]
  (map #(Character/digit % 10) (str num)))

(defn- required-sum [num]
  (let [d (digits num) n (count d)]
    (reduce + (map #(Math/pow % n) d))))

(defn armstrong? [num]
  (== num (required-sum num)))
